mono/metadata/icall.c

```diff
@@ -6056,23 +6061,43 @@ ves_icall_System_Reflection_RuntimeAssembly_GetForwardedTypes (MonoReflectionAss
 
 		mono_assembly_load_reference (image, assembly_idx - 1);
 		g_assert (image->references [assembly_idx - 1]);
-		if (image->references [assembly_idx - 1] == (gpointer)-1)
+		if (image->references [assembly_idx - 1] == (gpointer) - 1) {
+			MonoExceptionHandle ex = MONO_HANDLE_NEW (MonoException, mono_get_exception_bad_image_format ("Invalid image"));
+			MONO_HANDLE_ARRAY_SETREF (types, aindex, NULL_HANDLE);
+			MONO_HANDLE_ARRAY_SETREF (exceptions, aindex, ex);
+			exception_count++; aindex++;
 			continue;
+		}
 		MonoClass *klass = mono_class_from_name_checked (image->references [assembly_idx - 1]->image, nspace, name, error);
-		if (!is_ok (error))
+		if (!is_ok (error)) {
+			MonoExceptionHandle ex = mono_error_convert_to_exception_handle (error);
```

Comment:
So there's a problem here in that it leaves `error` in an unusuable state (by design).  (Same in other places where you do `if (!is_ok (error)) { ... convert ... continue;}`)

In this case we _know_ we're not going to propagate the error out.  We want to capture it and store it.

In order to do that, a better approach is to make a new  local`ERROR_DECL (captured_error)` inside the loop body and pass that to all these functions.

```
   for (...) {
      ERROR_DECL (captured_error);

      do_something (..., captured_error);
      if (!is_ok (captured_error)) {
            MonoExceptionHandle ex = mono_error_convert_to_exception_handle (captured_error);
            MONO_HANDLE_ARRAY_SETREF (exceptions, aindex, ex);
            aindex++;
            continue;
      }
      ...
  }
```